algorithms/src/sorting/impl/Kokkos_SortByKeyImpl.hpp

```diff
@@ -0,0 +1,319 @@
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values) {
+  if (keys.stride(0) == 1 && values.stride(0) == 1) {
+    sort_by_key_onedpl(space, keys, values);
+  } else {
+    sort_by_key_via_binsort(space, keys, values);
+  }
+}
+#endif
+
+// fallback case
+template <class ExecutionSpace, class KeysDataType, class... KeysProperties,
+          class ValuesDataType, class... ValuesProperties>
+std::enable_if_t<Kokkos::is_execution_space<ExecutionSpace>::value>
+sort_by_key_device_view_without_comparator(
+    const ExecutionSpace& space,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values) {
+  sort_by_key_via_binsort(space, keys, values);
+}
+
+// ---------------------------------------------------
+//
+// specialize cases for sorting by key with comparator
+//
+// ---------------------------------------------------
+
+#if defined(KOKKOS_ENABLE_CUDA)
+template <class ComparatorType, class KeysDataType, class... KeysProperties,
+          class ValuesDataType, class... ValuesProperties>
+void sort_by_key_device_view_with_comparator(
+    const Kokkos::Cuda& space,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values,
+    const ComparatorType& comparator) {
+  sort_by_key_cudathrust(space, keys, values, comparator);
+}
+#endif
+
+#if defined(KOKKOS_ENABLE_ROCTHRUST)
+template <class ComparatorType, class KeysDataType, class... KeysProperties,
+          class ValuesDataType, class... ValuesProperties>
+void sort_by_key_device_view_with_comparator(
+    const Kokkos::HIP& space,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values,
+    const ComparatorType& comparator) {
+  sort_by_key_hipthrust(space, keys, values, comparator);
+}
+#endif
+
+#if defined(KOKKOS_ENABLE_ONEDPL)
+template <class ComparatorType, class DataType, class... Properties>
+void sort_by_key_device_view_with_comparator(
+    const Kokkos::Experimental::SYCL& space,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values,
+    const ComparatorType& comparator) {
+  if (keys.stride(0) == 1 && values.stride(0) == 1) {
+    sort_by_key_onedpl(space, keys, values, comparator);
+  } else {
+    // FIXME
+    // copy_to_host_run_stdsort_copy_back(space, view);
+  }
+}
+#endif
+
+// fallback case
+template <class ComparatorType, class ExecutionSpace, class KeysDataType,
+          class... KeysProperties, class ValuesDataType,
+          class... ValuesProperties>
+std::enable_if_t<Kokkos::is_execution_space<ExecutionSpace>::value>
+sort_by_key_device_view_with_comparator(
+    const ExecutionSpace& space,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values,
+    const ComparatorType& comparator) {
+  // FIXME What do we do here?
```

Comment:
Not 100% sure what to do here. My guess is to do something like this on the host side
```
  auto zipped_begin =
      oneapi::dpl::make_zip_iterator(keys_host.data(), values_host.data());
  oneapi::dpl::sort(
      policy, zipped_begin, zipped_begin + n,
      [](auto lhs, auto rhs) { return comparator(std::get<0>(lhs), std::get<0>(rhs)); });

```
However, zip iterators are not available in C++. So either need to switch to pairs, or introduce one.

Reply:
Allocate a permutation array, initialize with iota, sort that array with user-defined functor that compares keys (`[keys](int i, int j) { returns compare(keys(i), keys(j); }` or so) , apply permutation to keys and values.

Reply:
Is there apply permutation utility anywhere in Kokkos?

Reply:
Not that I know of